vpr/src/pack/sync_netlists_to_routing_flat.cpp

```diff
@@ -0,0 +1,367 @@
+        AtomNetId atom_net_id = convert_to_atom_net_id(parent_net_id);
+
+        ClusterNetId clb_net_id;
+        for(auto& rt_node: tree->all_nodes()){
+            auto node_type = rr_graph.node_type(rt_node.inode);
+            if(node_type != IPIN && node_type != OPIN)
+                continue;
+
+            auto physical_tile = device_ctx.grid.get_physical_type({
+                rr_graph.node_xlow(rt_node.inode),
+                rr_graph.node_ylow(rt_node.inode),
+                rr_graph.node_layer(rt_node.inode)
+            });
+
+            int pin_index = rr_graph.node_pin_num(rt_node.inode);
+
+            auto [_, subtile] = get_sub_tile_from_pin_physical_num(physical_tile, pin_index);
+
+            ClusterBlockId clb = place_ctx.grid_blocks.block_at_location({
+                rr_graph.node_xlow(rt_node.inode),
+                rr_graph.node_ylow(rt_node.inode),
+                subtile,
+                rr_graph.node_layer(rt_node.inode)
+            });
+
+            if(!is_pin_on_tile(physical_tile, pin_index))
+                continue;
+
+            if(!clb_net_id){
+                clb_net_id = clb_netlist.create_net(atom_ctx.nlist.net_name(parent_net_id));
+                atom_lookup.set_atom_clb_net(atom_net_id, clb_net_id);
+            }
+
+            t_pb_graph_pin* pb_graph_pin = get_pb_graph_node_pin_from_block_pin(clb, pin_index);
+
+            ClusterPortId port_id = clb_netlist.find_port(clb, pb_graph_pin->port->name);
+            if(!port_id){
+                PortType port_type;
+                if(pb_graph_pin->port->is_clock)
+                    port_type = PortType::CLOCK;
+                else if(pb_graph_pin->port->type == IN_PORT)
+                    port_type = PortType::INPUT;
+                else if(pb_graph_pin->port->type == OUT_PORT)
+                    port_type = PortType::OUTPUT;
+                else
+                    VTR_ASSERT_MSG(false, "Unsupported port type");
+                port_id = clb_netlist.create_port(clb, pb_graph_pin->port->name, pb_graph_pin->port->num_pins, port_type);
+            }
+            PinType pin_type = node_type == OPIN ? PinType::DRIVER : PinType::SINK;
+
+            ClusterPinId new_pin = clb_netlist.create_pin(port_id, pb_graph_pin->pin_number, clb_net_id, pin_type, pb_graph_pin->pin_count_in_cluster);
+            clb_netlist.set_pin_net(new_pin, pin_type, clb_net_id);
+        }
+    }
+    /* 4. Rebuild internal cluster netlist lookups */
+    remapped = clb_netlist.compress();
+    rebuild_atom_nets_lookup(remapped);
+    /* 5. Rebuild place_ctx.physical_pins lookup
+     * TODO: maybe we don't need this fn and pin_index is enough? */
+    place_ctx.physical_pins.clear();
+    for(auto clb: clb_netlist.blocks()){
+        place_sync_external_block_connections(clb);
+    }
+    /* TODO: Remove rr_nodes added by the flat router */
+}
+
+/** Fix up pin rotation maps and the atom pin -> pb graph pin lookup for every block */
```

Comment:
Explain if this is only for primitives, and give LUTs as an example. (could put this in the Doxygen comment at the function declaration point if it is about how to use the function).